cmd/libs/go2idl/conversion-gen/generators/conversion.go

```diff
@@ -67,139 +80,125 @@ func DefaultNameSystem() string {
-			if fallbackPackage == t.Name.Package || !context.Universe.Package(fallbackPackage).Has(t.Name.Name) {
-				continue
-			}
-			return context.Universe.Package(fallbackPackage).Type(t.Name.Name), true
+func getPeerTypeFor(context *generator.Context, t *types.Type, potenialPeerPkgs []string) *types.Type {
+	for _, ppp := range potenialPeerPkgs {
+		p := context.Universe.Package(ppp)
+		if p == nil {
+			continue
+		}
+		if p.Has(t.Name.Name) {
+			return p.Type(t.Name.Name)
 		}
-		return nil, false
 	}
-	return context.Universe.Package(internalPackage).Type(t.Name.Name), true
+	return nil
 }
 
-type conversionType struct {
+type conversionPair struct {
 	inType  *types.Type
 	outType *types.Type
 }
 
 // All of the types in conversions map are of type "DeclarationOf" with
 // the underlying type being "Func".
-type conversions map[conversionType]*types.Type
+type conversionFuncMap map[conversionPair]*types.Type
 
-// Returns all already existing conversion functions that we are able to find.
-func existingConversionFunctions(context *generator.Context) conversions {
+// Returns all manually-defined conversion functions in the package.
+func getManualConversionFunctions(context *generator.Context, pkg *types.Package, manualMap conversionFuncMap) {
 	scopeName := types.Name{Package: conversionPackagePath, Name: "Scope"}
 	errorName := types.Name{Package: "", Name: "error"}
 	buffer := &bytes.Buffer{}
 	sw := generator.NewSnippetWriter(buffer, context, "$", "$")
 
-	preexisting := make(conversions)
-	for _, p := range context.Universe {
-		for _, f := range p.Functions {
-			if f.Underlying == nil || f.Underlying.Kind != types.Func {
-				glog.Errorf("Malformed function: %#v", f)
-				continue
-			}
-			if f.Underlying.Signature == nil {
-				glog.Errorf("Function without signature: %#v", f)
-				continue
-			}
-			signature := f.Underlying.Signature
-			// Check whether the function is conversion function.
-			// Note that all of them have signature:
-			// func Convert_inType_To_outType(inType, outType, conversion.Scope) error
-			if signature.Receiver != nil {
-				continue
-			}
-			if len(signature.Parameters) != 3 || signature.Parameters[2].Name != scopeName {
-				continue
-			}
-			if len(signature.Results) != 1 || signature.Results[0].Name != errorName {
-				continue
-			}
-			inType := signature.Parameters[0]
-			outType := signature.Parameters[1]
-			if inType.Kind != types.Pointer || outType.Kind != types.Pointer {
-				continue
-			}
-			// Now check if the name satisfies the convention.
-			args := argsFromType(inType.Elem, outType.Elem)
-			sw.Do("Convert_$.inType|public$_To_$.outType|public$", args)
-			if f.Name.Name == buffer.String() {
-				key := conversionType{inType.Elem, outType.Elem}
-				if v, ok := preexisting[key]; ok && v != nil {
-					panic(fmt.Sprintf("duplicate static conversion defined: %#v", key))
-				}
-				preexisting[key] = f
-			}
-			buffer.Reset()
+	for _, f := range pkg.Functions {
```

Comment:
A lot of individual commits are not the whole picture.  I tried to keep
each commit as a logical step towards the goal, but sometimes you need to
read ahead to know WHY something was done.  I can beef up some of the
per-commit comments.

Some of these COULD be broken up into distinct PRs, but I am loathe to do
so because many of them expose somewhat partially-complete internal
states.  Others build on each other such that extracting them would be more
rebase pain than I think it is worth - all 40 commits touch the same set of
about 8 files.

I think that you should be able to review commit-by-commit for all of the
go2idl and generator changes.  Only the Makefile itself is really best as a
whole final snapshot.

On Tue, Jul 5, 2016 at 4:16 AM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In cmd/libs/go2idl/conversion-gen/generators/conversion.go
> <https://github.com/kubernetes/kubernetes/pull/25978#discussion_r69545920>
> :
>
> > -			outType := signature.Parameters[1]
> > -			if inType.Kind != types.Pointer || outType.Kind != types.Pointer {
> > -				continue
> > -			}
> > -			// Now check if the name satisfies the convention.
> > -			args := argsFromType(inType.Elem, outType.Elem)
> > -			sw.Do("Convert_$.inType|public$_To_$.outType|public$", args)
> > -			if f.Name.Name == buffer.String() {
> > -				key := conversionType{inType.Elem, outType.Elem}
> > -				if v, ok := preexisting[key]; ok && v != nil {
> > -					panic(fmt.Sprintf("duplicate static conversion defined: %#v", key))
> > -				}
> > -				preexisting[key] = f
> > -			}
> > -			buffer.Reset()
> > +	for _, f := range pkg.Functions {
>
> @thockin <https://github.com/thockin> - I have a feeling that a lot of
> changes in this PR, is actually unrelated to not-checking-in the code.
> As an example, changes in this file.
>
> Would it be possible to split such things to separate PRs - it will take
> me ages to review this huge PR (I have a feeling that in my case review
> time is something like O(|PR|^3) :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25978/files/d4a5536797bea70a9ab71dd629fbcda0aa8d3254#r69545920>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFVgVF_G3LKg6pHkZJcELND_VExPCKffks5qSj0egaJpZM4Ijc3K>
> .
>